entity-framework/core/querying/user-defined-function-mapping.md

```diff
@@ -0,0 +1,154 @@
+FROM [Blogs] AS [b]
+WHERE [dbo].[CommentedPostCountForBlog]([b].[BlogId]) > 1
+```
+
+## Mapping a method to a custom SQL
+
+EF Core also allows for user-defined functions that get converted to a specific SQL. This can be done by specifying a SQL expression using the [Microsoft.EntityFrameworkCore.Query.SqlExpressions](/dotnet/api/microsoft.entityframeworkcore.query.sqlexpressions) API. The SQL expression is provided using `HasTranslation` method during user-defined function configuration.
+
+In the example below, we'll create a function that computes difference between two integers.
+
+The CLR method is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#HasTranslationFunctionDefinition)]
+
+The function definition is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#HasTranslationFunctionConfiguration)]
+
+Note that [SqlExpressionFactory](/dotnet/api/microsoft.entityframeworkcore.query.sqlexpressionfactory) is used to construct a `SqlExpression` tree.
+
+Once we define the function, it can be used in the query. Instead of calling database function, EF Core will translate the method body directly into SQL based on the SQL expression tree constructed from the HasTranslation. The following LINQ query:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Program.cs#HasTranslationQuery)]
+
+Produces the following SQL:
+
+```sql
+SELECT [p].[PostId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
+FROM [Posts] AS [p]
+WHERE [p].[PostId] < CASE
+    WHEN [p].[BlogId] > 3 THEN [p].[BlogId] - 3
+    ELSE 3 - [p].[BlogId]
+END
+```
+
+## Mapping a Queryable function to a table-valued gunction
+
+EF Core also supports mapping to a table-valued function using a user-defined CLR function returning an `IQueryable` of entity types, allowing EF Core to map TVFs with parameters. The process is similar to mapping a scalar user-defined function to a SQL function: we need a TVF in the database, a CLR function that is used in the LINQ queries, and a mapping between the two.
+
+As an example, we'll use a table-valued function that returns all posts having at least one comment that meets a given "Like" threshold:
+
+```sql
+CREATE FUNCTION dbo.PostsWithPopularComments(@likeThreshold int)
+RETURNS @posts TABLE
+(
+    PostId int not null,
+    BlogId int not null,
+    Content nvarchar(max),
+    Rating int not null,
+    Title nvarchar(max)
+)
+AS
+BEGIN
+    INSERT INTO @posts
+    SELECT[p].[PostId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
+    FROM[Posts] AS[p]
+    WHERE(
+        SELECT COUNT(*)
+        FROM[Comments] AS[c]
+        WHERE([p].[PostId] = [c].[PostId]) AND([c].[Likes] >= @likeThreshold)) > 0
+
+    RETURN
+END
+```
+
+The CLR function signature is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#QueryableFunctionDefinition)]
+
+> [!TIP]
```

Comment:
It can be used to start an EF Core query. Avoid query root. Use self-bootstrapping.